### PR TITLE
feat(gateway): Add gRPC job stream API

### LINKED ISSUE #11708: Add gRPC job stream API

### PR DESCRIPTION
**Pull request by berkaycanbc**
## Description
**Should be reviewed after https://github.com/camunda/zeebe/pull/13350.**

A new gRPC endpoint is defined to register clients to job streams. The new endpoint uses
ClientStreamAdapter to register and de-register clients when the connection is cancelled
or closed. GatewayTest and StubbedGateway classes are modified to handle gRPC server streaming
on test classes.

## Related issues

closes #11708 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by berkaycanbc** [comment]
@koevskinikola this PR is now ready for review! 🚀

File: gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    streamObserver.cancel();
+
+    // then
+    Awaitility.await("until the stream is removed")
+        .until(() -> !jobStreamer.containsStreamFor(jobType));
+  }
+
+  // TODO add test for onClose
+  // How to test onClose? it seems the graceful way to close a stream from the client side is
+  // simply to cancel it, as per the gRPC docs. So is onClose only for when we close it server side?
```
Comment (koevskinikola) [inline gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java:127]: :thinking: From my understanding, `onClose` is called when the stream is completed gracefully or with an error. So you already test that with the `shouldRemoveStreamOnError()` test case. 

You can also add `requestStream.onCompleted();` in the `TestStreamObserver#onCompleted()` method, and add an additional test case for completing the stream with the `streamObserver` instance.

//cc @npepinpe are my assumptions correct?

File: gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java
Comment (koevskinikola) [inline gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java:1]: :question: @berkaycanbc @npepinpe should we also include IT tests with this PR? Or do you plan to do it in another PR?

**Review by koevskinikola — CHANGES_REQUESTED** [review]
@berkaycanbc the runtime code looks great, nice job! :tada: 

I only have some suggestions for the test coverage.

File: gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    asyncClient.streamActivatedJobs(request, streamObserver);
+    jobStreamer.waitStreamToBeAvailable(BufferUtil.wrapString(jobType));
+
+    return streamObserver;
+  }
+
+  private static class TestStreamObserver
+      implements ClientResponseObserver<StreamActivatedJobsRequest, ActivatedJob> {
+    private ClientCallStreamObserver<StreamActivatedJobsRequest> requestStream;
+    private final List<ActivatedJob> streamedJobs;
```
[the remark this answers is not in the document]
Comment (berkaycanbc) [reply gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java:153]: Private fields of inner class and enclosing classes are interchangable but I agree that it is not a good practice to access them without getters. Updated! 👍

File: gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    jobStreamer.push(activatedJob).join();
+
+    // then
+    final ActivatedJob streamedActivatedJob = streamedJobs.get(0);
+    assertThat(streamedActivatedJob.getType()).isEqualTo(activatedJob.jobRecord().getType());
+    assertThat(activatedJob.jobRecord().getVariables()).isEqualTo(fetchedVariables);
+  }
+
+  @Test
+  public void shouldPushNoJobsWhenNonAvailable() {
```
[the remark this answers is not in the document]
Comment (berkaycanbc) [reply gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java:61]: I think what you suggested is definitely better. Updated! 👍

File: gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+  @Test
+  public void shouldMapRequestAndResponse() {
+    // given
+    final String jobType = "testJob";
+    final String worker = "testWorker";
+    final Duration timeout = Duration.ofMinutes(1);
+    final List<String> fetchVariables = Arrays.asList("foo", "bar");
+
+    final List<ActivatedJob> streamedJobs =
+        getStreamActivatedJobsRequest(jobType, worker, timeout, fetchVariables).streamedJobs;
```
[the remark this answers is not in the document]
Comment (berkaycanbc) [reply gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java:40]: Same as I commented above. 👍

File: gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java (hunk 1 above)
Comment (berkaycanbc) [reply gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java:127]: @koevskinikola Current test cases only trigger `onCancelHandler` hook. It is still not clear how we can trigger `onCloseHandler`. I think, it is better if we create an issue for that and have a separate PR (as we did for validations). **What do yo think?**

AFAIK, there isn't a way to set hooks for handling `onCompleted`. Based on the design, we need to remove the worker from the stream. The only way to de-register streams is to cancel the RPC call. Therefore, we decided not to have a case where client calls `onCompleted`. I think, client implementation will apply the same design principle.

Cc: @npepinpe

File: gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java
Comment (berkaycanbc) [reply gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java:1]: @koevskinikola Don't we need the client implementation first? I can create a new issue for that one as well. What do you think?

**Comment by berkaycanbc** [comment]
@koevskinikola Very useful comments! 🚀 It is now ready for a re-review.

File: gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java (hunk 1 above)
Comment (koevskinikola) [reply gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java:129]: I guess we should cover this case in the IT tests then. It would be good if you note this in the issue you create for them.

**Review by koevskinikola — APPROVED** [review]
The changes look good now @berkaycanbc :rocket:
